Python/Module2_EssentialsOfPython/Errors_and_Debugging.md

```diff
@@ -0,0 +1,112 @@
+
+<!-- #endregion -->
+
+<!-- #region -->
+## Exceptions
+An exception occurs when a piece of code is valid and understood by the interpreter, but for some reason the interpreter is unable to execute the action. Exceptions can occur when a function is passed an incompatible type of input, when the program is asked to something it cannot, and in many other circumstances. Python has many types of built-in exceptions that can be raised, but here are several that you are likely to encounter and their causes:
+
+| Exception  | Causes |
+| ------------- |:-------------:|
+| `IndexError` | Attempting to index or slice into a list with an index outside of the list's range |   
+| `KeyError` | Attempting to access a dictionary value with a key that is not in the dictionary |   
+| `NameError` | Referencing a variable or function that is not yet defined |   
+| `TypeError` | Passing an object of an invalid type into a function or operation |   
+| `ValueError` | Passing an object of an appropriate type but invalid value to a function or operation |
+| `ZeroDivisionError` | Attempting to divide by zero. Can also be raised when a function is passed a value that would force it to divide by zero |
+
+These six exceptions are the most common exceptions you will face, as they cover many possible issues. There exist many more specific exceptions for complicated errors, but they are much rarer and tend to only appear in very specific cases. Let's look at a piece of code that will raise an exception.
+
+```python
+# attempting to take the square root of a negative number
+>>> from math import sqrt
+>>> math.sqrt(-1)
+ValueError: math domain error
+
+```
+
+The math library cannot take the square root of a negative number. Because the number passed as an argument is technically the correct type, but is not compatible with the function, a `ValueError` is raised. Whenever you encounter one of these exceptions, be sure that you understand why the encounter is raised before you try to fix it.
+
+It is also possible to raise exceptions on purpose with the `raise` statement. If you are writing a function that is designed only for certain inputs, it can be worthwhile to verify that inputs will be valid. As an example, let's create our own square root function.
+
+```python
+def squareRoot(x):
+  if x < 0:
+    raise ValueError('Cannot take the square root of a negative number.')
+  return x**0.5
+  
+>>> print(squareRoot(-1))
+ValueError: Cannot take the square root of a negative number.
+```
+
+When the invalid value is passed, our function raises a ValueError rather than allowing the function to run with the value, possibly causing a worse or more confusing error. Clearly, in this case it was much better to raise an exception than to allow the program to continue! The `assert` statement can also be used to raise errors in the format `assert [boolean statement], 'error message'`, but it tends to be less informative and less powerful than simply using `raise`.
```

Comment:
I'm going to totally replace this with a complete section about assertions, since it doesn't make much sense in its current form. I think there's better discussion to be had about assertions than what I wrote here